Expression tests and related issues

wti opened the pull request:
As promised in #299, I filled in some of the expression tests, raising three kinds of issues:


**Test helper**
- The top-level method `actExp(..)` is to replace XCTest assertion directly.
It identifies the first text delta and prints expected and actual cleanly for copy/paste.  But it is relatively ungainly and not isolated, and there are likely diffing libraries that would do a better job.  It can merge and work as-is, but I can clean it up per your preferences.  I used it only in the new tests.

**Three kinds of test issues**

1.  3 tests fail due to JSON encoding differences, presumed due to locale differences
- I flagged these as expected fails. In the expected locale, these should be reported as unexpected pass.
- I can investigate whether JSONEncoder configuration will avoid locale-dependent behavior.

2.  `testAsyncExpression3()`: fails when there is no network
- It's unclear if we should check and skip, or use a different async call that does not require a network.

3. The test for "identifier expression" now refers to a quoted keyword.  Let me know if something else was intended.

**Two implementation issues**
1. optional string has quotes around the entire optional: `"Optional(\"0\")"`
2. inout variables are rendered constant (unusable) by macro expansion, resulting in compile errors.

**Merging**
Commits with aspects to investigate are marked DRAFT.  Commits are designed to merge even if the issues above require changes (e.g., tests should continue to pass; the current behavior is accepted for now).  Commits are also separated to isolate the issues above, to allow skipping a particular issue/commit.  The goal was to enable the changes to be merged, with follow-on work scheduled in later tasks.

wti commented:
Sorry, I can't see the logs for the failing Bitrise or Test checks.  I tested on macOS 14 with 5.9.  (It fails as expected on Linux due to missing Foundation API's and objc support.)

kishikawakatsumi commented:
Thank you for the PR! I'll check back later.

wti commented:
Sorry, please ignore the first commit to add `.project` to the `.gitignore` (cb6d4e...).  (I've added that to my global configuration.)

wti commented:
Please let me know if you'd like me to rebase or merge the branch.

kishikawakatsumi commented:
No worries. I just can't take my time.

kishikawakatsumi commented:
Thank you very much. It looks good to me. All tests are passing in my local environment, so if there is a problem when I run it on CI, I will fix it on my end.

wti commented on Tests/AssertTests.swift:
```diff
@@ -1023,6 +1024,7 @@ final class AssertTests: XCTestCase {
   }
 
   func testTryExpression2() throws {
+    XCTExpectFailure("99 vs 116 json encoded count d/t locale difference")
```
Sorry - Removed

wti commented on Tests/AssertTests.swift:
```diff
@@ -5907,6 +5909,7 @@ final class AssertTests: XCTestCase {
 
   @available(macOS 13.0, *)
   func testRegexLiteral() {
+    XCTExpectFailure("99 vs 116 String.Index d/t locale difference")
```
Sorry - Removed

wti commented:
Thank you for your patience.

I rebased off main, removed the unnecessary commit, and removed the expected failures per request.

fyi, four tests in AssertTests are failing for me:
- testMagicLiteralExpression2
- testInitializerExpression
- testSelectorExpression1
- testSelectorExpression2

(XCode 15.0, swift 5.9.0, M1 Studio)

If I discover a solution for inconsistent tests, I'll make a separate PR.

Thanks!
